// nodegrass/lib/nodegrass.js
/*!
 * Http-Https Request - nodegrass
 * Copyright(c) 2012 Sk ScottKiss
 * MIT Licensed
 */

/**
 * Module dependencies.
 */
var http = require('http'),
	 https = require('https'),
	 querystring = require('querystring'),
	 iconv = require('iconv-lite');
	 


function NodeGrass(){

	this.reqheaders = {
		'Accept':'text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5',
		'Accept-Encoding':'utf-8',
		'Accept-Language':'zh-cn,zh;q=0.8,en-us;q=0.5,en;q=0.3',
		'Cache-Control':'no-cache',
		'Connection':'keep-alive',
		'Content-Type':'text/xml; charset=utf-8',
		'User-Agent':	'Mozilla/5.0 (Windows NT 6.1; rv:14.0) Gecko/20100101 Firefox/14.0.1',
		//'X-Requested-With':'XMLHttpRequest'
	};
	this.reqtimeout = 10000;
	this.restimeout = 60000;
}


//Get Method Request
//Support HTTP and HTTPS request,and Automatic recognition
//@Param url
//@Param callback
NodeGrass.prototype.get = function(url,callback,charset){
	 var protocol = getProtocol(url);
	 var _defaultCharSet = 'utf8';
	 var content = "";
	 if(typeof charset === 'string' ){
			_defaultCharSet = charset;
	 }
	// console.log(_defaultCharSet);
	 if(protocol === 'http'){
		var options={
			host:getHost(url),
			port:80,
			path:getPath(url),
			method:'GET',
			headers:this.reqheaders
		};
		var request_timer = null, req = null;

		// 请求5秒超时
		request_timer = setTimeout(function() {
			req.abort();
			console.log('Request Timeout.');
		}, this.reqtimeout);


		req = http.request(options,function(res){
			clearTimeout(request_timer);
			// 等待响应60秒超时
			var response_timer = setTimeout(function() {
				res.destroy();
				console.log('Response Timeout.');
			}, this.restimeout);

			var status = res.statusCode;
			var headers = res.headers;
			if(_defaultCharSet==="gbk"){
				res.setEncoding('binary');
			}else{
				res.setEncoding(_defaultCharSet);
			}
			res.on('data',function(chunk){
				content+=chunk;
			});
			res.on('end',function(){
				clearTimeout(response_timer);
				if(_defaultCharSet==="gbk"){
					content = iconv.decode(new Buffer(content,'binary'),'gbk');
				}
				callback(content,status,headers);
			});

			res.on('error', function(e) {
				// 响应头有错误
				clearTimeout(request_timer);
				console.log("Got error: " + e.message);
			});


		});
		req.write("\n");
		//req.setSocketKeepAlive(true);
		req.end();
		return req;
		
	 }else if(protocol === 'https'){
		var options={
			host:getHost(url),
			port:443,
			path:getPath(url),
			method:'GET',
			headers:this.reqheaders
		};

		var request_timer = null, req = null;

		// 请求5秒超时
		request_timer = setTimeout(function() {
			req.abort();
			console.log('Request Timeout.');
		}, this.reqtimeout);
		
		req = https.request(options,function(res){
			clearTimeout(request_timer);
			// 等待响应60秒超时
			var response_timer = setTimeout(function() {
				res.destroy();
				console.log('Response Timeout.');
			}, this.restimeout);
			var status = res.statusCode;
			var headers = res.headers;
			if(_defaultCharSet==="gbk"){
				res.setEncoding('binary');
			}else{
				res.setEncoding(_defaultCharSet);
			}
			res.on('data',function(chunk){
				content+=chunk;
			});
			res.on('end',function(){
				clearTimeout(response_timer);
				if(_defaultCharSet==="gbk"){
					content = iconv.decode(new Buffer(content,'binary'),'gbk');
				}
				callback(content,status,headers);
			});

			res.on('error', function(e) {
				// 响应头有错误
				clearTimeout(request_timer);
				console.log("Got error: " + e.message);
			});
		});
		req.write("\n");
		req.end();
		return req;


		}else{
		throw "sorry,this protocol do not support now";
	 }

}

//Post Method Request
//Support HTTP and HTTPS request,and Automatic recognition
//@Param url
//@Param callback
//@Param header
//@param postdata
NodeGrass.prototype.post = function(url,callback,reqheaders,data,charset){
	var protocol = getProtocol(url);
	 var _defaultCharSet = 'utf8';
	 var content = "";
	 if(typeof charset === 'string' ){
			_defaultCharSet = charset;
	 }
	 reqheaders = reqheaders || this.reqheaders;
	console.log(reqheaders);
	data = querystring.stringify(data);
	//console.log(data);
	if(protocol==='http'){
		var options={
			host:getHost(url),
			port:80,
			path:getPath(url),
			method:'POST',
			headers:reqheaders
		};
		var request_timer = null, req = null;

		// 请求5秒超时
		request_timer = setTimeout(function() {
			req.abort();
			console.log('Request Timeout.');
		}, this.reqtimeout);

		req = http.request(options,function(res){
			clearTimeout(request_timer);
			// 等待响应60秒超时
			var response_timer = setTimeout(function() {
				res.destroy();
				console.log('Response Timeout.');
			}, this.restimeout);
			var status = res.statusCode;
			var headers = res.headers;
			if(_defaultCharSet==="gbk"){
				res.setEncoding('binary');
			}else{
				res.setEncoding(_defaultCharSet);
			}
			res.on('data',function(chunk){
				content+=chunk;
			});
			res.on('end',function(){
				clearTimeout(response_timer);
				if(_defaultCharSet==="gbk"){
					content = iconv.decode(new Buffer(content,'binary'),'gbk');
				}
				callback(content,status,headers);
			});
			res.on('error', function(e) {
				// 响应头有错误
				clearTimeout(request_timer);
				console.log("Got error: " + e.message);
			});
		});
		req.write(data+"\n");
		req.end();
		return req;
		
		
	 }else if(protocol==='https'){
		var options={
			host:getHost(url),
			port:443,
			path:getPath(url),
			method:'POST',
			headers:reqheaders
		};

		var request_timer = null, req = null;
		// 请求5秒超时
		request_timer = setTimeout(function() {
			req.abort();
			console.log('Request Timeout.');
		}, this.reqtimeout);

		var req = https.request(options,function(res){
			clearTimeout(request_timer);
			// 等待响应60秒超时
			var response_timer = setTimeout(function() {
				res.destroy();
				console.log('Response Timeout.');
			}, this.restimeout);
			var status = res.statusCode;
			var headers = res.headers;
			if(_defaultCharSet==="gbk"){
				res.setEncoding('binary');
			}else{
				res.setEncoding(_defaultCharSet);
			}
			res.on('data',function(chunk){
				content+=chunk;
			});
			res.on('end',function(){
				clearTimeout(response_timer);
				if(_defaultCharSet==="gbk"){
					content = iconv.decode(new Buffer(content,'binary'),'gbk');
				}
				callback(content,status,headers);
			});
			res.on('error', function(e) {
				// 响应头有错误
				clearTimeout(request_timer);
				console.log("Got error: " + e.message);
			});		
		});
		req.write(data+"\n");
		req.end();
		return req;
	
		}else{
		throw "sorry,this protocol do not support now";
	 }

}

//Parse the url,get the path
//e.g. http://www.google.com/path/another -> /path/another
function getPath(url){
	var pathPattern = /\w+:\/\/([^\/]+)(\/.+)(\/$)?/i;
	var fullPath = url.match(pathPattern);
	return fullPath?fullPath[2]:'/';
}

//Parse the url,get the host name
//e.g. http://www.google.com/path/another -> www.google.com
function getHost(url){
	var hostPattern = /\w+:\/\/([^\/]+)(\/)?/i;
	var domain = url.match(hostPattern);
	//console.log(domain[1]);
	return domain[1];
}

//Get the Protocol
//http://www.google.com/path/another => http
function getProtocol(url){
	return url.substring(0,url.indexOf(":"));
}
var nodegrass = new NodeGrass();
nodegrass.NodeGrass = NodeGrass;
module.exports = nodegrass;
